torch/testing/_internal/common_subclass.py

```diff
@@ -0,0 +1,259 @@
+import operator
+import torch
+from copy import deepcopy
+from functools import reduce
+from torch.utils._pytree import tree_map
+
+# TODO: Move LoggingTensor here.
+from torch.testing._internal.logging_tensor import LoggingTensor
+
+
+# Base class for wrapper-style tensors.
+class WrapperTensor(torch.Tensor):
```

Comment:
not sure I'm too happy about all the flagrant copy pasting. Do you really need the fully fledged subclasses here to adequately test?

Reply:
The idea here is to maintain a set of semi-realistic reference subclasses that cover the various cases we expect and can be used throughout testing. I think these will become more valuable as the subclass work continues and testing across torch.nn is expanded to be subclass-aware. If there's a better place for these subclass definitions, I'm happy to move them, but I think it's important that they're made available eventually, and this PR seemed like a decent place to land an initial version of them.

Reply:
I feel like a good start would be to route the duplicates into these implementations.

Reply:
I changed my mind: you shouldn't test with realistic reference subclasses (except for regression tests for things we specifically found in the wild). Instead, I'd like it quite a bit better if we hand constructed some artificial subclasses that covered the edge cases for the current implementation.

Reply:
I agree with you in theory, but I think it's pretty hard to identify what the edge cases are a priori. In contrast, the testing strategy currently taken ensures all the normal things users expect to be able to do with parameters work correctly.

In fact, the current tensor subclass test set almost does what you're saying, as each exercises things differently:
* A simple wrapper tensor (`LoggingTensor` already exists in `torch/testing/_internal/logging_tensor.py` so I just used it)
* A simple non-wrapper tensor containing extra state (the fairly artificial `NonWrapperTensor`)
* A tensor that reports a different shape than its underlying storage (e.g. `DiagTensorBelow`)
* A tensor that contains multiple tensor component but presents a single-tensor interface (`SparseTensor`)